### ISSUE TITLE
rainbow-delimeters face same as identifier face

### ISSUE THREAD

**Issue by stevemolitor** [protesilaos/ef-themes#28]
I'm not sure if this is fixable or not, but I noticed that in clojure-mode, often `raindbow-delimeters-depth-3-face` is the same color as the function call name:

![](https://cdn.zappy.app/133211b7f09557c8109a57f8dd8b3916.png)

I like rainbow-delimeters in lisp, but in this case the `(` bleeds into `println` function name, making things less clear.

**Comment by stevemolitor** [comment]
> Taking a step back, this is a general problem with the rainbow delimiters: some parenthesis will look similar to text in its immediate context.

Yeah that's what I figured. I could tweak the ordering of the colors to fix this problem, but it might cause the same problem elsewhere - a game of whack-a-mole.

Thanks for the explanation.

**Comment by protesilaos** [comment]
This problem will persist because we do not know ahead of time the context of each delimiter.  One solution is to use colours that are not common in the theme, but this is likely to make the overall appearance too busy/colourful.  Beside that, it limits the range of possible hues, which will beat the purpose of easily identifying distinct levels of nesting.

Maybe the method of `highlight-parentheses` will be more viable?  It only colourises delimiters in the immediate context.  We can set it up to work with the Ef themes.

**Comment by stevemolitor** [comment]
Thanks for the suggestion @protesilaos. Personally I'm not a fan of `highlight-parentheses`. I like seeing parens colored even when my cursor is not in the expression. When it works, `rainbow-delimiters-mode` helps me read code anywhere in the visible portion of my buffer. I don't care for the "flickering" effect of `highlight-parentheses`. Using `highlight-parentheses` might work well for others though. It's kind of a personal preference thing.

You've helped clarify the issue with `rainbow-delimiters-mode` for me though, thanks. I'm starting to think the best solution would involve a tree-sitter powered rainbow delimiters mode. Or at least a mode that can detect the face of the thing adjacent to a paren. If the rainbow mode wants to color something `rainbow-3`, but the adjacent thing has the same foreground as `rainbow-3`, then use `rainbow-7` instead. If `rainbow-7`, then `rainbow-3`. Iow `((n + 4) % 8) - 1`. 

Maybe I'll write such a mode someday. Or maybe I won't - no promises! But yeah for now this is a fundamental limitation of rainbow delimeters that ef-themes can't fix.

**Comment by mok0** [comment]
Would it be possible to make available separate color mappings for org mode headings? I notice that it's not possible to override the `ef-themes-heading-*`. I like the rainbox-delimiters to be colorful and display an actual rainbow, but not the Org mode headings.

**Comment by protesilaos** [comment]
Hello @mok0!

> Would it be possible to make available separate color mappings for org mode headings?

Yes, I will make it happen for the next stable version.

**Comment by protesilaos** [comment]
By the way, I am closing this issue because there is nothing we can do to address the original topic.
